### PR TITLE
feat(ui): G10.3-T3 bulk targets.yaml import — paste/upload → preview → in-process create/update CRUD

### LINKED ISSUE #875: Task: G10.3-T3 Bulk targets.yaml import UI — paste/upload → server-side preview (create/update) → confirm via in-process create/update CRUD
Parent initiative: #340
Parent goal: #336
Depends-on: #873 (the bulk-import preview renders into the targets list)

## Target

- Add: `/ui/connectors/import` (parse+preview) + confirm routes + Jinja2 preview-table template.
- Add: `backend/tests/test_ui_connectors_import.py`.
- Precedent: #873 list (`ui/routes/connectors/list_view.py`); T2 #874 in-process REST
  delegation pattern (`ui/routes/connectors/forms.py` calls `create_target` / `update_target`
  from `meho_backplane.api.v1.targets` directly — mirror this for the confirm step); the
  `meho targets import` CLI tool (G0.3-T6 [#257](https://github.com/evoila/meho/issues/257))
  whose **client-orchestrated CRUD** classification logic this UI re-implements server-side;
  the REST CRUD routes it calls — `POST /api/v1/targets` + `PATCH /api/v1/targets/{name}`
  (G0.3-T3 [#254](https://github.com/evoila/meho/issues/254)); chassis CSRF (#866).

## Context

Per #340 work-item #5 — the operator pastes or uploads a `targets.yaml` → server-side parse
(`yaml.safe_load`) → HTMX preview table classifying each entry CREATE-vs-UPDATE → confirm →
the route applies the plan **in-process** via the existing target CRUD (`create_target` for
new names, `update_target` for existing ones), exactly as the `meho targets import` CLI does
client-side (parse → `GET` existing names → classify → `POST` new / `PATCH` existing).
**There is no `/api/v1/targets/import` server endpoint** — the CLI tool (#257) orchestrates
plain CRUD, and this UI mirrors that orchestration. `tenant_admin` only.

Mapping rules (parity with `cli/internal/cmd/targets/import.go`):
- Known top-level keys (`name`, `product`, `host`, `port`, `auth_model`, `secret_ref`,
  `vpn_required`, `aliases`, `notes`, `extras`, …) map 1:1 to `TargetCreate` / `TargetUpdate` fields.
- Unknown keys spill into the `extras` JSONB column.
- `fingerprint` is server-managed (probe-only writer) → dropped from import with a preview warning.
- On UPDATE, emit a sparse body (only keys present in the YAML); `name`/`product` are not patched.

## Current state

The list view (#873) + create/edit forms (#874) exist, but bulk target import is CLI-only
(`meho targets import`, #257). No web UI for it.

## Desired state

A paste box / file upload accepts a `targets.yaml`; server-side parse renders a preview table
marking each target CREATE-vs-UPDATE (HTMX); parse errors render inline; confirm applies the
plan via in-process `create_target` / `update_target` and shows a result summary (N created,
M updated).

## Acceptance criteria

- [ ] Paste or upload of a `targets.yaml` renders a server-side preview table marking each
      target CREATE-vs-UPDATE (HTMX); YAML parse errors render inline (no 500).
- [ ] Confirm applies the plan in-process via `POST /api/v1/targets` (new) + `PATCH
      /api/v1/targets/{name}` (existing); a result summary (N created, M updated) is shown.
- [ ] Classification + key-mapping match `meho targets import` (#257): known keys → columns,
      unknown → `extras`, `fingerprint` dropped, CREATE vs UPDATE decided by existing-name lookup.
- [ ] `tenant_admin` only (server-side 403 for operators, via `resolve_operator_or_403`);
      CSRF enforced; cross-tenant isolation (import only into the caller's tenant; `resolve_target`).
- [ ] Goal #336 **G10.3** checklist line ticked; `ruff` + `mypy` clean;
      `pytest -n auto backend/tests/test_ui_connectors_import.py` passes.

## Out of scope

- List/detail — #873. Create/edit forms — #874. A dedicated `/api/v1/targets/import` REST
  endpoint — not in scope; if one is ever wanted it's a separate G0.3 backend task, but the
  CLI tool (#257) deliberately uses plain CRUD and this UI follows suit.

## References

- Precedent: #873 list (`ui/routes/connectors/list_view.py`); #874 forms
  (`ui/routes/connectors/forms.py` — in-process `create_target`/`update_target` delegation);
  CLI orchestration `cli/internal/cmd/targets/import.go` (G0.3-T6 #257); REST CRUD routes
  G0.3-T3 [#254](https://github.com/evoila/meho/issues/254); [`docs/cross-repo/targets-yaml.md`](../blob/main/docs/cross-repo/targets-yaml.md).
- HTMX: https://htmx.org
- Best-practices anchors (by pack name): **frontend_best_practices** — HTMX upload+preview+confirm, CSRF; **python_best_practices** — FastAPI multipart/form parse, `yaml.safe_load`, in-process CRUD reuse, RBAC.

### PR DESCRIPTION
**Pull request by zdamir**
## Summary
- Adds the bulk `targets.yaml` import UI at `/ui/connectors/import` (work item #5 of Initiative #340 G10.3): paste OR upload a `targets.yaml` → server-side `yaml.safe_load` parse → HTMX preview table classifying each entry CREATE-vs-UPDATE → confirm → apply the plan **in-process** via the existing target CRUD (`create_target` for new names, `update_target` for existing).
- **No `/api/v1/targets/import` endpoint** — this mirrors the client-orchestrated CRUD the `meho targets import` CLI (#257) performs. `import_view`'s key-mapping + classification is a server-side port of `import.go`'s `mapEntry` / `buildLivePlan`, so the web import and CLI import produce byte-identical writes: known keys → columns, unknown → `extras` JSONB (merged with an explicit `extras:` block), `fingerprint` dropped with a warning, UPDATE emits a sparse body (`name`/`product` stripped) so re-imports don't wipe omitted columns.
- Preview→confirm is stateless: confirm re-parses + re-classifies against the tenant's current targets (a target created between preview and confirm is PATCHed, not re-CREATEd into a 409). `tenant_admin`-only via `resolve_operator_or_403`, CSRF-gated by the chassis middleware, cross-tenant-isolated at both the existing-name lookup and the in-process CRUD layer. Parse errors render inline (422, never a 500).
- Regenerated the CLI OpenAPI snapshot + `client.gen.go` for the new `/ui/connectors/import` routes; CLI builds clean.

## Test plan
- [x] `ruff check` / `ruff format --check` (connectors package + test) — clean
- [x] `mypy src/meho_backplane/ui/routes/connectors/` — clean (9 files)
- [x] `pytest -n auto backend/tests/test_ui_connectors_import.py` — 19 passed
- [x] `pytest tests/test_ui_connectors_{import,forms,view}.py` — 65 passed (no T1/T2 regression)
- [x] `code-quality.py --diff` — 0 blockers (2 sub-threshold warnings)
- [x] `cli && make snapshot-openapi && make generate && go build ./...` — new routes present, client compiles
- [x] Paste + upload preview classify CREATE/UPDATE; malformed YAML + missing required field render inline (no 500)
- [x] Confirm creates new + sparse-PATCHes existing; sparse PATCH preserves omitted columns; malformed YAML on confirm writes nothing
- [x] `tenant_admin` only (operator 403); missing CSRF 403; cross-tenant isolation (import lands in caller tenant only; same-name in another tenant does not flip CREATE→UPDATE)

## Out of scope
- List/detail (#873), create/edit forms (#874). A dedicated `/api/v1/targets/import` REST endpoint — deliberately not added; the CLI (#257) uses plain CRUD and this UI follows suit.
- `import_view.py` is 479 lines / `submit_confirm` 81 lines — both under the 600/100 block thresholds (sub-threshold code-quality warnings).

Closes #875

---

## Follow-up (auto-implement-issue, iteration 1, 2026-05-26)

Addressed the REQUEST_CHANGES review findings from `/auto-review-pr` iteration 1:

- **B1** `42d9f17` — Validate the full import plan before any write. `build_plan` now constructs and schema-validates every `TargetCreate`/`TargetUpdate` body up front (catching Pydantic `ValidationError` + the `AuthModel` enum `ValueError`) and raises `ImportParseError` before the confirm write loop runs. A structurally-valid YAML with a schema-invalid value (bad `auth_model` enum, out-of-range `port`) now renders the inline 422 fragment and writes nothing — no more partial import + 500. The same validation runs in `render_preview`. Regression tests cover preview-inline + confirm-422-with-zero-writes (including the valid row ahead of the bad entry).

Adjacent (non-blocking, fixed while in `import_router.py`):

- Corrected the `_YAML_MAX_BYTES` comment — a multipart upload is bounded by Starlette `MultiPartParser.max_part_size` (1 MiB default), not the CSRF middleware's urlencoded-only body cap.

Disputed / not addressed (per the review's own disposition):

- CodeRabbit majors on `cli/api/openapi.json` declaring the 422 response as `text/html` — a generic FastAPI codegen artifact for any body-param route, not a PR defect; hand-editing the deterministic snapshot would fail the freshness check.
- CodeRabbit nitpick on markdown heading style in `docs/codebase/ui.md` — cosmetic.

## Summary by CodeRabbit

* **New Features**
  * Added bulk import functionality for connector targets via YAML files or text paste.
  * Includes a preview step to review create/update actions before confirming.
  * Accessible to tenant admins via a new "Bulk import" link in the connectors page.

* **Documentation**
  * Updated codebase documentation with bulk import feature details and workflow.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/1181?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by zdamir** [comment]
@coderabbitai review

File: backend/src/meho_backplane/ui/routes/connectors/import_view.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+    existing = await _existing_names(db_session, session_ctx.tenant_id)
+    plan = build_plan(entries, existing)
+
+    created = 0
+    updated = 0
+    for entry in plan:
+        if entry.action == "CREATE":
+            await create_target(
+                body=_build_create_body(entry.body),
+                operator=operator,
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/routes/connectors/import_view.py:515]: **[🛑 B1]** — schema-invalid entry 500s mid-write and leaves a partial import

**What:** `submit_confirm` constructs the Pydantic write body inside the mutating loop — `_build_create_body(entry.body)` / `_build_update_body(entry.body)` are called per-iteration, immediately before `create_target` / `update_target`. `build_plan` / `render_preview` only validate parse-level required fields (`name`/`product`/`host`); they never construct `TargetCreate`/`TargetUpdate`. So a YAML that parses cleanly but carries a schema-invalid value — e.g. `auth_model: NOT_A_REAL_ENUM` (raises `ValueError` in `_build_create_body`'s `AuthModel(...)` coercion), an out-of-range `port`, etc. — passes the preview, then raises mid-loop on confirm.

**Why it matters:** Two regressions. (1) The unhandled `ValueError`/`ValidationError` surfaces as an HTTP 500 — breaking the acceptance criterion "parse errors render inline (no 500)". (2) Entries earlier in the list were already written via `create_target`/`update_target` before the failing entry, so the tenant is left with a **partial import** and no rollback. This diverges from the CLI parity contract this UI is explicitly mirroring: `import.go` builds the *entire* plan (`buildLivePlan`) before `executePlan` issues the first write ("no partial write — the plan is built before any API call fires"). Confirmed: `_build_create_body({...,"auth_model":"NOT_A_VALID_ENUM"})` raises `ValueError: 'NOT_A_VALID_ENUM' is not a valid AuthModel`.

**Required change:** Build and validate every `TargetCreate`/`TargetUpdate` for the whole plan *before* entering the write loop, catching `pydantic.ValidationError` and the `ValueError` from the `AuthModel(...)` coercion. On any validation failure, render the inline 422 preview-error fragment (same shape as the `ImportParseError` path) and write nothing. Apply the same pre-validation in `render_preview` so the preview itself flags a schema-invalid entry instead of showing a clean plan that then fails on confirm. Add a test: a structurally-valid YAML with an invalid `auth_model` (or out-of-range field) asserting HTTP 422, the inline-error fragment, and zero rows written.

**Source:** `cli/internal/cmd/targets/import.go` lines 229–271 (plan built before any write); `backend/src/meho_backplane/ui/routes/connectors/import_view.py:443-458`. (Also flagged by CodeRabbit as 🟠 Major.)

**Review by zdamir — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #875 — G10.3-T3 bulk `targets.yaml` import UI (paste/upload → preview → in-process create/update CRUD)
**Surface area:** backend (FastAPI + HTMX/Jinja2) + generated CLI snapshot
**Best-practices packs audited:** python, frontend
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `5cb4fc6`
**AI reviewer (CodeRabbit):** received in 368s

### Acceptance criteria
- [x] Paste/upload renders a server-side CREATE-vs-UPDATE preview (HTMX) — verified `test_preview_paste_renders_create_and_update_table`, `test_preview_upload_file_renders_table`.
- [⚠️] YAML parse errors render inline (no 500) — **partially met**. YAML *syntax* errors + missing-required-field render inline 422 (`test_preview_malformed_yaml_renders_inline_error_not_500`). A structurally-valid YAML that fails **schema** validation (bad `auth_model` enum, out-of-range `port`, …) is NOT caught at preview and **500s mid-write on confirm** (see 🛑 B1).
- [x] Confirm applies plan in-process via `create_target` (new) + `update_target` (existing); summary shown — verified `test_confirm_creates_and_updates_in_process`. Call signatures match `api/v1/targets.py`.
- [x] Classification/key-mapping match `meho targets import` (#257): known→columns, unknown→`extras`, `fingerprint` dropped, CREATE/UPDATE by existing-name lookup — verified against `cli/internal/cmd/targets/import.go` `mapEntry`/`buildLivePlan` line-by-line; matching unit tests pass.
- [x] `tenant_admin` only via `resolve_operator_or_403` (operators 403); CSRF enforced; cross-tenant isolation — verified `test_import_*_rejects_operator_role_with_403`, `test_import_preview_rejects_missing_csrf`, `test_confirm_imports_only_into_callers_tenant`, `test_preview_classification_scoped_to_caller_tenant`. Tenant comes from `operator.tenant_id` server-side; body cannot override.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #875 | In Review | Closes | ✅ | ⚠️ (one AC partial) | fix B1, then close on merge |

### Predicted vs. actual diff
- ✅ `import_router.py` (thin routes), `import_view.py` (parse/classify/confirm), `import.html` / `_import_preview.html` / `_import_result.html`, `connectors/__init__.py` (router include before detail), `list.html` (Bulk import link), `test_ui_connectors_import.py`, `docs/codebase/ui.md`, `cli/api/openapi.json` + `client.gen.go` (regenerated).
- ⚠️ `import_view.py` `submit_confirm` — expected but wrong: Pydantic body construction deferred into the write loop → partial-write + 500 (see inline B1).
- ❓ None unexpected.
- 🚫 None undisclosed.
- 🕳️ None missing.

### Integration verification
- [x] `ruff check` (connectors pkg + test) — clean
- [x] `ruff format --check` — 10 files already formatted
- [x] `mypy src/.../connectors/` — clean (9 files)
- [x] `pytest test_ui_connectors_import.py` — 19 passed
- [x] `pytest test_ui_connectors_{forms,view}.py` — 46 passed (no T1/T2 regression)
- [x] `cd cli && make snapshot-openapi && make generate` — **zero diff** (CLI API snapshot freshness GREEN)
- [x] CLI parity audit vs `import.go` `mapEntry`/`buildLivePlan`/`parseTargetsYAML` — known-key set, `_SKIP_SILENT`={fingerprint}, extras-merge, sparse UPDATE (`name`/`product` stripped), soft-delete-excluded existing-name lookup all match.
- [x] DCO — green; route-registration order (`build_import_router` before `build_detail_router`) — verified; `yaml.safe_load` (not `yaml.load`) — verified; upload bounded by Starlette `MultiPartParser.max_part_size=1MB` (422 on exceed), paste bounded by `Form(max_length=256KiB)`.

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 B1 — `submit_confirm` builds `TargetCreate`/`TargetUpdate` inside the mutating write loop; a schema-invalid entry (e.g. bad `auth_model`) raises mid-loop → unhandled 500 + partial import (rows before it already written). Breaks AC "parse errors render inline (no 500)" and the CLI's no-partial-write contract. Required change: build + validate the full plan (catch `pydantic.ValidationError` / `ValueError` from the enum coercion) *before* the write loop; on failure render the inline 422 preview error and write nothing. Mirror in `render_preview` so validation errors surface at preview time too. Add a test for a structurally-valid-but-schema-invalid entry asserting 422 + zero writes. *(Also flagged by CodeRabbit as 🟠 Major.)*

### Adjacent improvements (non-blocking, not posted inline)
- `import_router.py` docstring (lines 55–58) attributes the upload body bound to "the chassis CSRF middleware … caps the buffered body at 256 KiB". That cap only runs for `application/x-www-form-urlencoded` (`csrf.py` `_resolve_presented_token` returns early for other content types); the multipart upload is actually bounded by Starlette's `MultiPartParser.max_part_size` (1 MiB default, 422 on exceed). Behaviour is safe; the comment's stated mechanism is inaccurate.
- CodeRabbit nitpick: the "Bulk import" doc subsections use `**bold**` instead of `####` headings, inconsistent with the Memory/Connectors sections in the same file. Cosmetic.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 — the partial-write/500 finding (B1), confirmed against the code and the missing test.
- Disputed / out of scope: 2 — both `cli/api/openapi.json` findings ("declare 422 as `text/html`"). The `422` JSON shape is FastAPI's auto-generated validation-error schema injected for every body-param route regardless of `response_class=HTMLResponse`; it is a generic codegen artifact affecting all `/ui/*` POST routes, not a defect this PR introduced. Hand-editing the generated snapshot to add `text/html` would fail the deterministic `CLI API snapshot freshness` check. Not actionable here.
- Nitpick: 1 — markdown heading style (recorded under Adjacent improvements).

### Docs
- [x] `docs/codebase/ui.md` — updated: full "Bulk import (Task #875)" section (mapping rules, parity, statelessness, RBAC/CSRF/isolation, inline-422 parse errors, route table, file inventory). Accurate.

### What worked well (review-body only)
- CLI parity is genuine and faithful — `_map_entry`/`_entry_to_update_body`/`_parse_targets_yaml` are a line-accurate port of `import.go`, including the extras pre-pass merge and the `name`/`product` strip on UPDATE.
- Cross-tenant isolation is tested at both layers (existing-name lookup scoped to tenant; same-name-in-other-tenant does not flip CREATE→UPDATE), and the sparse-PATCH test asserts omitted columns survive — honest contract tests, not happy-path.
- CLI snapshot regenerated proactively; freshness check green on first pass (the recurring iter-1 blocker is cleared here).

**Comment by zdamir** [comment]
@coderabbitai review

**Review by zdamir — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #875 — bulk `targets.yaml` import UI (paste/upload → preview → in-process create/update CRUD)
**Surface area:** backend (FastAPI/HTMX UI) + docs
**Best-practices packs audited:** python
**Verdict:** APPROVE (posted as COMMENT-state — GitHub blocks an explicit APPROVE on a self-authored PR; the binding verdict is `APPROVE` in the result file)
**Head SHA reviewed:** `42d9f17`
**AI reviewer (CodeRabbit):** re-review not received within window — fix worker re-posted `@coderabbitai review` (ack 20:48:06Z); no new review submitted after ~14 min wait. Iter-1 CodeRabbit findings already cross-checked and carried forward.

### Iteration-1 blocker resolution
- **🛑 B1 (partial-import / unhandled 500 on schema-invalid entry) — RESOLVED.**
  - `build_plan` now constructs **and** schema-validates every `TargetCreate` / `TargetUpdate` body up front (catching `pydantic.ValidationError` + the `AuthModel` enum `ValueError`) and raises `ImportParseError` *before* `submit_confirm`'s write loop. The write loop now consumes the pre-validated `entry.body` directly — no per-entry Pydantic construction can raise mid-write.
  - `render_preview` runs the identical `build_plan` path, so the preview can no longer green-light a plan that confirm would reject (verified: both routes compute `existing`, then `_parse_targets_yaml` + `build_plan` inside one try/except, both returning `_render_inline_error` on `ImportParseError`).
  - The three inline-422 sites are collapsed into the shared `_render_inline_error` helper (DRY, single fail-fast render path).
  - Write-loop branch keys on `isinstance(entry.body, TargetCreate)`; `TargetCreate` and `TargetUpdate` are independent `BaseModel` subclasses (neither inherits the other), so CREATE/UPDATE routing is unambiguous.

### Acceptance criteria
- [x] Parse/validation errors render inline (no 500) — schema-invalid value surfaces inline 422 on **preview** — `test_preview_schema_invalid_value_renders_inline_error` (422 + `data-import-error`)
- [x] No partial import on schema-invalid YAML — confirm 422s and writes nothing, incl. a **valid row positioned before the bad entry** — `test_confirm_schema_invalid_entry_422_writes_nothing` asserts `_count_targets == 0` and `_load_target("valid-first") is None`
- [x] `build_plan` validates bodies up front (CLI no-partial-write parity) — `test_build_plan_raises_on_schema_invalid_auth_model` + `test_build_plan_raises_on_out_of_range_port`

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #875 | In Review | Closes | ✅ | ✅ full | close on merge |

### Predicted vs. actual diff (delta `5cb4fc6..42d9f17`)
- ✅ `import_view.py` — validation lifted into `build_plan`; shared `_render_inline_error`; `PlanEntry.body` now a validated model + new `.fields` projection
- ✅ `import_router.py` — corrected `_YAML_MAX_BYTES` comment (Starlette `MultiPartParser.max_part_size`, not the CSRF urlencoded-only cap)
- ✅ `tests/test_ui_connectors_import.py` — +4 regression tests (19 → 23); pre-existing tests correctly updated for the `body`-is-model / `.fields` change
- ✅ `docs/codebase/ui.md` — no-partial-write contract documented
- 🚫 none undisclosed
- 🕳️ none missing

### Integration verification
- [x] `ruff check` (connectors routes + import test) — clean
- [x] `ruff format --check` — 3 files already formatted
- [x] `mypy src/.../connectors/ --ignore-missing-imports` — no issues, 9 source files
- [x] `pytest tests/test_ui_connectors_import.py` — **23 passed** (faithful `uv sync --frozen` venv)
- [x] `pytest tests/test_ui_connectors_{forms,view}.py` — 46 passed (no T1/T2 regression)
- [x] CI: Python (ruff+mypy+pytest) SUCCESS, Python integration testcontainers SUCCESS, Go SUCCESS, **CLI API snapshot freshness SUCCESS** (undisturbed by the fix)
- [⚠️ infra] CI "Build backplane image" FAILURE — Docker Hub `429 Too Many Requests` (unauthenticated pull rate limit) resolving the `python` base image; transient registry flake, independent of the diff — not a blocker
- migrations replay / docker compose / agent eval — n/a (no schema, route, migration, or agent path touched in this delta)

### Findings
None. The single iteration-1 blocker is resolved and proven by regression tests; the fresh blast-radius pass on the delta surfaced no new issues.

### Adjacent improvements (non-blocking, not posted inline)
- CodeRabbit nitpick on `docs/codebase/ui.md`: "Bulk import" subsections use `**bold**` instead of `####` headings, inconsistent with the Memory/Connectors sections. Cosmetic; deferred.
- CodeRabbit majors on `cli/api/openapi.json` declaring the 422 response as `text/html`: FastAPI's auto-generated validation-error schema artifact for any body-param route regardless of `response_class=HTMLResponse`; hand-editing the deterministic snapshot would break the CLI-snapshot-freshness gate. Disputed (carried forward from iter-1).

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 (B1 partial-write/500 — fixed this iteration)
- Disputed: 2 (openapi 422 content-type codegen artifacts)
- Out of scope: 1 (markdown heading nitpick)
- Re-review of `42d9f17`: not received within the wait window.

### Docs
- [x] `docs/codebase/ui.md` — updated (no-partial-write contract + "validate the whole plan before any write" section; preview-validates-identically note)

### What worked well
- The no-partial-write regression test is the right test: a valid row positioned *before* the schema-invalid entry, asserting zero rows committed — that is the exact contract, not a happy-path proxy.
- Validation and rendering were unified cleanly: `build_plan` is the single validation seam and `_render_inline_error` the single fail-fast render path, eliminating the preview/confirm divergence risk by construction.
